File: python/src/onemin/resources/text.py
"""Text domain resource for the 1min.ai API."""
from __future__ import annotations
from typing import Any, AsyncGenerator, Generator, TYPE_CHECKING

from onemin.resources._base_resource import BaseResource
from onemin.models import TextResult
from onemin._streaming import stream_sse
from onemin._constants import API_KEY_HEADER

if TYPE_CHECKING:
    from onemin._base_client import BaseOneMinClient


class TextResource(BaseResource):
    """Text generation and LLM chat operations. Domain timeout: 30s."""

    _domain = "text"

    def chat(
        self,
        prompt: str,
        *,
        model: str = "gpt-4o",
        stream: bool = False,
        web_search: bool = False,
        num_of_site: int | None = None,
        max_word: int | None = None,
        **kwargs: Any,
    ) -> TextResult | Generator[str, None, None]:
        """Send a chat/completion request to a language model.

        Hits POST /api/chat-with-ai with type ``UNIFY_CHAT_WITH_AI`` -- the
        unified endpoint that replaces the legacy CHAT_WITH_IMAGE,
        CHAT_WITH_PDF, and CHAT_WITH_YOUTUBE_VIDEO feature types. Pass
        attachments via ``kwargs['attachments'] = {"images": [...], "files": [...]}``
        and history settings via ``kwargs['settings']`` to access the full
        promptObject schema documented at
        https://docs.1min.ai/docs/api/chat-with-ai-api.

        Args:
            prompt: The user message to send. Including a YouTube URL in the
                prompt triggers automatic transcript extraction (max 3 URLs).
            model: Model name (e.g. 'gpt-4o', 'claude-3-5-sonnet', 'gemini-1.5-pro').
            stream: If True, returns a generator yielding token strings via SSE.
            web_search: Enable web search augmentation.
            num_of_site: Optional override for the number of sites to query
                (only applied when web_search is True; default upstream is 3).
            max_word: Optional cap on words pulled from each site
                (only applied when web_search is True; default upstream is 1000).
            **kwargs: Extra fields merged into promptObject. Useful keys:
                ``conversationId``, ``attachments``, ``settings``,
                ``brandVoiceId``, ``metadata``.

        Returns:
            TextResult (non-streaming) or Generator[str] (streaming).

        Example:
            result = client.text.chat("What is 2+2?")
            print(result.content)
        """
        prompt_object: dict[str, Any] = {"prompt": prompt}
        if web_search:
            web_search_settings: dict[str, Any] = {"webSearch": True}
            if num_of_site is not None:
                web_search_settings["numOfSite"] = num_of_site
            if max_word is not None:
                web_search_settings["maxWord"] = max_word
            prompt_object["settings"] = {"webSearchSettings": web_search_settings}
        prompt_object.update(kwargs)

        payload: dict[str, Any] = {
            "type": "UNIFY_CHAT_WITH_AI",
            "model": model,
            "promptObject": prompt_object,
        }

        if stream:
            return self._stream_chat(payload)

        response = self._client._request(
            "POST", "/api/chat-with-ai", json=payload, timeout=self._timeout,
        )
        return self._parse_text_result(response, model)

    def _stream_chat(self, payload: dict[str, Any]) -> Generator[str, None, None]:
        """Internal: stream chat via SSE."""
        url = f"{self._client._base_url}/api/chat-with-ai?isStreaming=true"
        headers = {API_KEY_HEADER: self._client._api_key}
        yield from stream_sse(
            self._client._client,
            url,
            headers=headers,
            json=payload,
            timeout=self._timeout,
        )

    async def achat(
        self,
        prompt: str,
        *,
        model: str = "gpt-4o",
        stream: bool = False,
        web_search: bool = False,
        num_of_site: int | None = None,
        max_word: int | None = None,
        **kwargs: Any,
    ) -> TextResult | AsyncGenerator[str, None]:
        """Send a chat/completion request to a language model asynchronously.

        Args:
            prompt: The user message to send.
            model: Model name (e.g. 'gpt-4o', 'claude-3-5-sonnet', 'gemini-1.5-pro').
                See ``Models.Text`` for all supported model names.
            stream: If True, returns an async generator yielding token strings via SSE.
            web_search: Enable web search augmentation.
            num_of_site: Optional override for the number of sites to query
                (only applied when web_search is True).
            max_word: Optional cap on words pulled from each site
                (only applied when web_search is True).
            **kwargs: Extra fields merged into promptObject (e.g. ``conversationId``,
                ``attachments``, ``settings``, ``brandVoiceId``, ``metadata``).

        Returns:
            TextResult (non-streaming) or AsyncGenerator[str] (streaming).

        Example:
            result = await client.text.achat("What is 2+2?")
            print(result.content)
        """
        prompt_object: dict[str, Any] = {"prompt": prompt}
        if web_search:
            web_search_settings: dict[str, Any] = {"webSearch": True}
            if num_of_site is not None:
                web_search_settings["numOfSite"] = num_of_site
            if max_word is not None:
                web_search_settings["maxWord"] = max_word
            prompt_object["settings"] = {"webSearchSettings": web_search_settings}
        prompt_object.update(kwargs)

        payload: dict[str, Any] = {
            "type": "UNIFY_CHAT_WITH_AI",
            "model": model,
            "promptObject": prompt_object,
        }

        if stream:
            return self._astream_chat(payload)

        response = await self._client._request(
            "POST", "/api/chat-with-ai", json=payload, timeout=self._timeout,
        )
        return self._parse_text_result(response, model)

    async def _astream_chat(self, payload: dict[str, Any]) -> AsyncGenerator[str, None]:
        """Internal: stream chat via async SSE."""
        from onemin._streaming import astream_sse
        url = f"{self._client._base_url}/api/chat-with-ai?isStreaming=true"
        headers = {API_KEY_HEADER: self._client._api_key}
        async for token in astream_sse(
            self._client._http,
            url,
            headers=headers,
            json=payload,
            timeout=self._timeout,
        ):
            yield token

    @staticmethod
    def _parse_text_result(response: dict[str, Any], model: str) -> TextResult:
        """Extract TextResult from API response.

        The /api/chat-with-ai endpoint nests the result at
        ``aiRecord.aiRecordDetail.resultObject`` (a list of strings). Older
        responses placed it at ``aiRecord.resultObject`` as a list or as a
        dict with ``message``/``content``/``text``. All shapes are handled.
        """
        ai_record = response.get("aiRecord", {})
        detail = ai_record.get("aiRecordDetail") or {}
        result_obj = detail.get("resultObject")
        if result_obj is None:
            result_obj = ai_record.get("resultObject")
        if isinstance(result_obj, list):
            content = "".join(str(c) for c in result_obj)
        elif isinstance(result_obj, dict):
            content = (
                result_obj.get("message")
                or result_obj.get("content")
                or result_obj.get("text")
                or str(result_obj)
            )
        else:
            content = "" if result_obj is None else str(result_obj)
        return TextResult(content=content, model=model)

File: typescript/src/resources/text.ts
import { BaseResource } from './base-resource.js';
import type { BaseOneMinClient } from '../base-client.js';
import type { TextResult } from '../types.js';
import { streamSSE } from '../streaming.js';
import { API_KEY_HEADER } from '../constants.js';

export interface ChatOptions {
  model?: string;
  stream?: boolean;
  webSearch?: boolean;
  /** Number of sites web search should query (only applied when webSearch is true). */
  numOfSite?: number;
  /** Maximum words extracted per site (only applied when webSearch is true). */
  maxWord?: number;
  [key: string]: unknown;
}

export class TextResource extends BaseResource {
  constructor(client: BaseOneMinClient) {
    super(client, 'text');
  }

  /**
   * Send a chat/completion request to a language model.
   *
   * Hits POST /api/chat-with-ai with type `UNIFY_CHAT_WITH_AI` -- the unified
   * endpoint that replaces the legacy CHAT_WITH_IMAGE / CHAT_WITH_PDF /
   * CHAT_WITH_YOUTUBE_VIDEO feature types. Pass `attachments`,
   * `conversationId`, `settings`, `brandVoiceId`, or `metadata` through
   * `options` to access the full promptObject schema documented at
   * https://docs.1min.ai/docs/api/chat-with-ai-api.
   *
   * @param prompt - The user message to send. Including a YouTube URL in the
   *   prompt triggers automatic transcript extraction (max 3 URLs).
   * @param options - Optional parameters: model (default "gpt-4o"), stream,
   *   webSearch, numOfSite, maxWord, plus any promptObject fields
   *   (`attachments`, `conversationId`, `settings`, etc.).
   * @returns TextResult with content, model name, and optional usage metadata.
   *   When stream is true, returns an AsyncGenerator yielding token strings.
   *
   * @example
   * const result = await client.text.chat('What is 2+2?');
   * console.log(result.content);
   *
   * @example
   * // Streaming
   * for await (const token of await client.text.chat('Tell me a joke', { stream: true })) {
   *   process.stdout.write(token);
   * }
   */
  async chat(prompt: string, options: ChatOptions & { stream: true }): Promise<AsyncGenerator<string, void, undefined>>;
  async chat(prompt: string, options?: ChatOptions & { stream?: false }): Promise<TextResult>;
  async chat(prompt: string, options?: ChatOptions): Promise<TextResult | AsyncGenerator<string, void, undefined>>;
  async chat(
    prompt: string,
    options: ChatOptions = {},
  ): Promise<TextResult | AsyncGenerator<string, void, undefined>> {
    const {
      model = 'gpt-4o',
      stream = false,
      webSearch = false,
      numOfSite,
      maxWord,
      ...extra
    } = options;

    const promptObject: Record<string, unknown> = { prompt, ...extra };
    if (webSearch) {
      const webSearchSettings: Record<string, unknown> = { webSearch: true };
      if (numOfSite !== undefined) webSearchSettings.numOfSite = numOfSite;
      if (maxWord !== undefined) webSearchSettings.maxWord = maxWord;
      promptObject.settings = { webSearchSettings };
    }

    const payload: Record<string, unknown> = {
      type: 'UNIFY_CHAT_WITH_AI',
      model,
      promptObject,
    };

    if (stream) {
      return this._streamChat(payload, model);
    }

    const response = await this.client._request<Record<string, unknown>>(
      'POST',
      '/api/chat-with-ai',
      payload,
      { timeout: this.timeout },
    );
    return this._parseTextResult(response, model);
  }

  private async _streamChat(
    payload: Record<string, unknown>,
    _model: string,
  ): Promise<AsyncGenerator<string, void, undefined>> {
    const url = `${this.client.baseUrl}/api/chat-with-ai?isStreaming=true`;
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          [API_KEY_HEADER]: (this.client as unknown as { apiKey: string }).apiKey,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
        signal: ctrl.signal,
      });
      clearTimeout(timer);
      return streamSSE(response);
    } catch (err) {
      clearTimeout(timer);
      throw err;
    }
  }

  private _parseTextResult(
    response: Record<string, unknown>,
    model: string,
  ): TextResult {
    const aiRecord = (response.aiRecord ?? {}) as Record<string, unknown>;
    const detail = (aiRecord.aiRecordDetail ?? {}) as Record<string, unknown>;
    const resultObj = detail.resultObject ?? aiRecord.resultObject;
    let content: string;
    if (Array.isArray(resultObj)) {
      content = resultObj.map((c) => String(c)).join('');
    } else if (resultObj && typeof resultObj === 'object') {
      const r = resultObj as Record<string, unknown>;
      content = String(r.message ?? r.content ?? r.text ?? JSON.stringify(r));
    } else {
      content = resultObj == null ? '' : String(resultObj);
    }
    return { content, model };
  }
}
